app/modules/dlc/services/RemoteFileService.scala

```diff
@@ -0,0 +1,38 @@
+package modules.dlc.services
+
+import javax.inject.Inject
+
+import modules.dlc.models.RemoteFile
+import play.api.Logger
+import play.api.libs.ws.WSClient
+
+import scala.concurrent.{ExecutionContext, Future}
+
+trait RemoteFileService {
+  def checkAvailability(file: RemoteFile): Future[Boolean]
+}
+
+class RemoteFileServiceFactory @Inject() (wsClient : WSClient)(implicit ec: ExecutionContext) {
+
+  private val logger = Logger(classOf[RemoteFileServiceFactory])
+
+  def checkAvailability(file: RemoteFile): Future[Boolean] = {
+    val fileService =  mapping.find { case(key,value) => file.url.startsWith(key) }.map(_._2)
+
+    fileService.map { service =>
+      service.checkAvailability(file)
+    } getOrElse {
+      logger.error(s"unsupported file hoster for url: ${file.url}")
+      Future.failed(new UnsupportedFileHosterException)
+    }
+  }
+
+  def mapping = {
+    Map(
+      "http://www.share-online.biz" -> new ShareOnlineFileService(wsClient),
```

Comment:
you could provide an keyMethod within the service (interface) then you could inject them all with mutibinder and register them...

your map would be like  allService.map(_.getUrl, this)